src/libfsm/minimise.c

```diff
@@ -7,49 +7,660 @@
+#include <sys/time.h>
+#endif
+
+#include "minimise_internal.h"
+
 int
 fsm_minimise(struct fsm *fsm)
 {
-	int r;
+	int r = 0;
+	struct fsm *dst = NULL;
 
-	assert(fsm != NULL);
+	unsigned char labels[256];
+	size_t label_count, orig_states, minimised_states;
+	fsm_state_t *mapping = NULL;
+
+#if LOG_TIME
+	struct timeval tv_pre, tv_post;
+
+#define TIME(T) if (0 != gettimeofday(&T, NULL)) { assert(0); }
+#define LOG_TIME_DELTA(NAME)				\
+	fprintf(stderr, "%-8s %.3f msec\n", NAME,	\
+	    1000.0 * (tv_post.tv_sec - tv_pre.tv_sec)	\
+	    + (tv_post.tv_usec - tv_pre.tv_usec)/1000.0);
+#else
+#define TIME(T)
+#define LOG_TIME_DELTA(NAME)
+#endif
 
 	/* This should only be called with a DFA. */
+	assert(fsm != NULL);
 	assert(fsm_all(fsm, fsm_isdfa));
 
-	/*
-	 * Brzozowski's algorithm.
-	 */
-
-	r = fsm_reverse(fsm);
-	if (!r) {
+	/* The algorithm used below won't remove states without a path
+	 * to an end state, because it cannot prove they're
+	 * unnecessary, so they must be trimmed away first. */
+	if (fsm_trim(fsm, FSM_TRIM_START_AND_END_REACHABLE) < 0) {
 		return 0;
 	}
 
-	r = fsm_determinise(fsm);
-	if (!r) {
-		return 0;
+	if (fsm->statecount == 0) {
+		return 1;	/* empty -- no-op */
 	}
 
-	r = fsm_reverse(fsm);
-	if (!r) {
-		return 0;
+	TIME(tv_pre);
+	collect_labels(fsm, labels, &label_count);
+	TIME(tv_post);
+	LOG_TIME_DELTA("collect_labels");
+
+	if (label_count == 0) {
+		return 1;	/* no edges -- no-op */
+	}
+
+	mapping = f_malloc(fsm->opt->alloc,
+	    fsm->statecount * sizeof(mapping[0]));
+	if (mapping == NULL) {
+		goto cleanup;
 	}
 
-	r = fsm_determinise(fsm);
+	orig_states = fsm->statecount;
+
+	TIME(tv_pre);
+	r = build_minimised_mapping(fsm, labels, label_count,
+	    mapping, &minimised_states);
+	TIME(tv_post);
+	LOG_TIME_DELTA("minimise");
+
 	if (!r) {
-		return 0;
+		goto cleanup;
+	}
+
+	/* Minimisation should never add states. */
+	assert(minimised_states <= orig_states);
+
+	/* Use the mapping to consolidate the current states
+	 * into a new DFA, combining states that could not be
+	 * proven distinguishable. */
+	TIME(tv_pre);
+	dst = fsm_consolidate(fsm, mapping, fsm->statecount);
+	TIME(tv_post);
+	LOG_TIME_DELTA("consolidate");
+
+	if (dst == NULL) {
+		r = 0;
+		goto cleanup;
+	}
+
+	fsm_move(fsm, dst);
+
+cleanup:
+	if (mapping != NULL) {
+		f_free(fsm->opt->alloc, mapping);
 	}
 
-	return 1;
+	return r;
+}
+
+/* Build a bit set of labels used, then write the set
+ * into a sorted array. */
+static void
+collect_labels(const struct fsm *fsm,
+    unsigned char *labels, size_t *label_count)
+{
+	size_t count = 0;
+	uint64_t label_set[4] = { 0, 0, 0, 0 };
+	int i;
+
+	fsm_state_t id;
+	for (id = 0; id < fsm->statecount; id++) {
+		struct fsm_edge e;
+		struct edge_iter ei;
+		unsigned char label;
+		for (edge_set_reset(fsm->states[id].edges, &ei);
+		     edge_set_next(&ei, &e); ) {
+			assert(e.state < fsm->statecount);
+			label = e.symbol;
+
+			if (label_set[label/64] & (1UL << (label & 63))) {
```

Comment:
Where does 64 come from? is this `sizeof (fsm_state_t) * CHAR_BIT`?

Reply:
It's from the `uint64_t label_set[4]` defined above -- it's a bitset of FSM_SIGMA_COUNT entries. I'll change the type to `uint64_t label_set[FSM_SIGMA_COUNT/64]`, but I'm not sure what sort of note on the `/64` and `&63` part would make it clearer without visually obscuring the pattern. Any suggestions?

Reply:
Commented in bf06b2ff.

Reply:
Ok, that works for me :)